cmd/testrunner.go

```diff
@@ -156,15 +173,68 @@ func testTypeCommandActionFactory(runner testrunner.TestRunner) cobraext.Command
 			return fmt.Errorf("cannot determine if package has data streams: %w", err)
 		}
 
+		configFileFlag := ""
+		runSetup := false
+		runTearDown := false
+		runTestsOnly := false
+
+		if runner.CanRunSetupTeardownIndependent() && cmd.Flags().Lookup(cobraext.ConfigFileFlagName) != nil {
+			// not all test types define these flags
+			runSetup, _ = cmd.Flags().GetBool(cobraext.SetupFlagName)
+			runTearDown, _ = cmd.Flags().GetBool(cobraext.TearDownFlagName)
+			runTestsOnly, _ = cmd.Flags().GetBool(cobraext.NoProvisionFlagName)
+
+			configFileFlag, err = cmd.Flags().GetString(cobraext.ConfigFileFlagName)
+			if err != nil {
+				return cobraext.FlagParsingError(err, cobraext.ConfigFileFlagName)
+			}
+			if configFileFlag != "" {
+				absPath, err := filepath.Abs(configFileFlag)
+				if err != nil {
+					return fmt.Errorf("cannot obtain the absolute path for config file path: %s", configFileFlag)
+				}
+				if _, err := os.Stat(absPath); err != nil {
+					return fmt.Errorf("can't find config file %s: %w", configFileFlag, err)
+				}
+				configFileFlag = absPath
+			}
+		}
+
 		signal.Enable()
 
 		var testFolders []testrunner.TestFolder
 		if hasDataStreams && runner.CanRunPerDataStream() {
 			var dataStreams []string
-			// We check for the existence of the data streams flag before trying to
-			// parse it because if the root test command is run instead of one of the
-			// subcommands of test, the data streams flag will not be defined.
-			if cmd.Flags().Lookup(cobraext.DataStreamsFlagName) != nil {
+
+			if runner.CanRunSetupTeardownIndependent() && runSetup || runTearDown || runTestsOnly {
+				if runTearDown || runTestsOnly {
+					locationManager, err := locations.NewLocationManager()
+					if err != nil {
+						return fmt.Errorf("can't read configuration directory")
+					}
+
+					type setupData struct {
+						ConfigFilePath string `json:"config_file_path"`
+					}
+					var serviceSetupData setupData
+					setupDataPath := filepath.Join(locationManager.ServiceSetupDir(), testrunner.ServiceSetupDataFileName)
+					fmt.Printf("Reading service setup data from file: %s\n", setupDataPath)
+					contents, err := os.ReadFile(setupDataPath)
+					if err != nil {
+						return fmt.Errorf("failed to read service setup data %q: %w", setupDataPath, err)
+					}
+					err = json.Unmarshal(contents, &serviceSetupData)
+					if err != nil {
+						return fmt.Errorf("failed to decode service setup data %q: %w", setupDataPath, err)
+					}
+					configFileFlag = serviceSetupData.ConfigFilePath
+				}
+				dataStream := testrunner.ExtractDataStreamFromPath(configFileFlag, packageRootPath)
+				dataStreams = append(dataStreams, dataStream)
+			} else if cmd.Flags().Lookup(cobraext.DataStreamsFlagName) != nil {
```

Comment:
Should we mark this flag as mutually exclusive with the others?

Reply:
Added --data-streams flag to be mutually exclusive wit the flags `--setup`, `--no-provision` and `--tear-down`